### PR TITLE
CallTarget ByRef: Fix ByRef Generics arguments

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR fixes an issue on byref generic arguments. 

@DataDog/apm-dotnet

### CODE REVIEW

File: tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs [hunk 1]
```diff
@@ -132,6 +132,20 @@ internal static DynamicMethod CreateBeginMethodDelegate(Type integrationType, Ty
                         callGenericTypes.Add(parameterProxyType);
                     }
                 }
+                else if (targetParameterType.IsByRef && targetParameterType.GetElementType().IsGenericParameter)
+                {
+                    // ByRef generic parameters needs to be unwrapped before accessing the `IsGenericParameter` property.
+                    var genTargetParameterType = genericArgumentsTypes[targetParameterType.GetElementType().GenericParameterPosition];
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs:138]: Not sure if you prefer it, but can avoid calling `GetElementType()` twice if you like?
```suggestion
                else if (targetParameterType.IsByRef && targetParameterType.GetElementType() is { IsGenericParameter: true } elementType)
                {
                    // ByRef generic parameters needs to be unwrapped before accessing the `IsGenericParameter` property.
                    var genTargetParameterType = genericArgumentsTypes[elementType.GenericParameterPosition];
```

**Review by andrewlock — APPROVED** [review]
LGTM 👍

File: tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs (hunk 1 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs:138]: 🤯

**Comment by andrewlock** [comment]
## Benchmarks Report :snail:

Benchmarks for #2099 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9): 
* 2 benchmarks are **slower**, with geometric mean 1.137
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`WriteAndFlushEnrichedTraces`|net472|912.9&nbsp;μs|15.10&nbsp;μs|14.13&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|795.0&nbsp;μs|5.94&nbsp;μs|5.26&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
|#2099|`WriteAndFlushEnrichedTraces`|net472|940.6&nbsp;μs|13.27&nbsp;μs|11.76&nbsp;μs|0.0000| 0.0000|0.0000|3.09&nbsp;KB|
|#2099|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|741.8&nbsp;μs|14.18&nbsp;μs|13.26&nbsp;μs|0.0000| 0.0000|0.0000|2.51&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendRequest`|netcoreapp3.1|274,887.7659&nbsp;ns|4,796.2897&nbsp;ns|4,251.7859&nbsp;ns|0.2747| 0.0000|0.0000|19617&nbsp;B|
|#2099|`SendRequest`|net472|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|#2099|`SendRequest`|netcoreapp3.1|288,105.7025&nbsp;ns|5,703.6199&nbsp;ns|5,335.1696&nbsp;ns|0.1471| 0.0000|0.0000|19617&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`ExecuteNonQuery`|net472|328.2&nbsp;ns|3.67&nbsp;ns|3.07&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`ExecuteNonQuery`|netcoreapp3.1|286.3&nbsp;ns|5.66&nbsp;ns|5.29&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|#2099|`ExecuteNonQuery`|net472|337.1&nbsp;ns|6.51&nbsp;ns|9.34&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
|#2099|`ExecuteNonQuery`|netcoreapp3.1|290.5&nbsp;ns|3.80&nbsp;ns|3.56&nbsp;ns|0.0000| 0.0000|0.0000|0&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`CallElasticsearch`|net472|3.232&nbsp;μs|0.0572&nbsp;μs|0.0955&nbsp;μs|0.1286| 0.0000|0.0000|827&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`CallElasticsearch`|netcoreapp3.1|1.612&nbsp;μs|0.0300&nbsp;μs|0.0619&nbsp;μs|0.0110| 0.0000|0.0000|768&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`CallElasticsearchAsync`|net472|3.280&nbsp;μs|0.0613&nbsp;μs|0.0656&nbsp;μs|0.1502| 0.0000|0.0000|963&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`CallElasticsearchAsync`|netcoreapp3.1|1.789&nbsp;μs|0.0348&nbsp;μs|0.0427&nbsp;μs|0.0121| 0.0000|0.0000|888&nbsp;B|
|#2099|`CallElasticsearch`|net472|3.304&nbsp;μs|0.0649&nbsp;μs|0.0575&nbsp;μs|0.1275| 0.0000|0.0000|827&nbsp;B|
|#2099|`CallElasticsearch`|netcoreapp3.1|1.788&nbsp;μs|0.0137&nbsp;μs|0.0129&nbsp;μs|0.0107| 0.0000|0.0000|768&nbsp;B|
|#2099|`CallElasticsearchAsync`|net472|3.415&nbsp;μs|0.0633&nbsp;μs|0.0867&nbsp;μs|0.1505| 0.0000|0.0000|963&nbsp;B|
|#2099|`CallElasticsearchAsync`|netcoreapp3.1|1.694&nbsp;μs|0.0312&nbsp;μs|0.0276&nbsp;μs|0.0124| 0.0000|0.0000|888&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`ExecuteAsync`|net472|3.511&nbsp;μs|0.0685&nbsp;μs|0.1145&nbsp;μs|0.1683| 0.0000|0.0000|1083&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`ExecuteAsync`|netcoreapp3.1|1.869&nbsp;μs|0.0250&nbsp;μs|0.0234&nbsp;μs|0.0137| 0.0000|0.0000|1008&nbsp;B|
|#2099|`ExecuteAsync`|net472|3.462&nbsp;μs|0.0646&nbsp;μs|0.0572&nbsp;μs|0.1687| 0.0000|0.0000|1083&nbsp;B|
|#2099|`ExecuteAsync`|netcoreapp3.1|1.824&nbsp;μs|0.0290&nbsp;μs|0.0257&nbsp;μs|0.0136| 0.0000|0.0000|1008&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendAsync`|net472|6.089&nbsp;μs|0.1154&nbsp;μs|0.1023&nbsp;μs|0.3486| 0.0000|0.0000|2.19&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendAsync`|netcoreapp3.1|4.238&nbsp;μs|0.0815&nbsp;μs|0.0762&nbsp;μs|0.0298| 0.0000|0.0000|2.09&nbsp;KB|
|#2099|`SendAsync`|net472|5.953&nbsp;μs|0.1190&nbsp;μs|0.1169&nbsp;μs|0.3504| 0.0000|0.0000|2.19&nbsp;KB|
|#2099|`SendAsync`|netcoreapp3.1|4.241&nbsp;μs|0.0815&nbsp;μs|0.0762&nbsp;μs|0.0295| 0.0000|0.0000|2.09&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|net472|3.815&nbsp;μs|0.0694&nbsp;μs|0.0649&nbsp;μs|0.2872| 0.0000|0.0000|1.79&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|netcoreapp3.1|3.631&nbsp;μs|0.0555&nbsp;μs|0.0519&nbsp;μs|0.0272| 0.0000|0.0000|1.94&nbsp;KB|
|#2099|`EnrichedLog`|net472|4.001&nbsp;μs|0.0747&nbsp;μs|0.0971&nbsp;μs|0.2871| 0.0000|0.0000|1.79&nbsp;KB|
|#2099|`EnrichedLog`|netcoreapp3.1|3.691&nbsp;μs|0.0576&nbsp;μs|0.0539&nbsp;μs|0.0270| 0.0000|0.0000|1.94&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|net472|170.7&nbsp;μs|2.07&nbsp;μs|1.84&nbsp;μs|0.6868| 0.2576|0.0000|5.23&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|netcoreapp3.1|151.1&nbsp;μs|2.12&nbsp;μs|1.77&nbsp;μs|0.0754| 0.0000|0.0000|5.05&nbsp;KB|
|#2099|`EnrichedLog`|net472|176.6&nbsp;μs|3.43&nbsp;μs|3.37&nbsp;μs|0.7082| 0.2656|0.0000|5.23&nbsp;KB|
|#2099|`EnrichedLog`|netcoreapp3.1|152.7&nbsp;μs|2.57&nbsp;μs|3.06&nbsp;μs|0.0000| 0.0000|0.0000|5.05&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|net472|11.62&nbsp;μs|0.231&nbsp;μs|0.367&nbsp;μs|0.8555| 0.0000|0.0000|5.31&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|netcoreapp3.1|10.11&nbsp;μs|0.198&nbsp;μs|0.175&nbsp;μs|0.0883| 0.0000|0.0000|6.28&nbsp;KB|
|#2099|`EnrichedLog`|net472|11.91&nbsp;μs|0.207&nbsp;μs|0.323&nbsp;μs|0.8541| 0.0000|0.0000|5.31&nbsp;KB|
|#2099|`EnrichedLog`|netcoreapp3.1|10.08&nbsp;μs|0.142&nbsp;μs|0.133&nbsp;μs|0.0932| 0.0000|0.0000|6.28&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #2099

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.RedisBenchmark.SendReceive&#8209;net472 | 1.144 | 2,144.05 | 2,453.12 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendReceive`|net472|2.141&nbsp;μs|0.0279&nbsp;μs|0.0261&nbsp;μs|0.1549| 0.0000|0.0000|987&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`SendReceive`|netcoreapp3.1|1.955&nbsp;μs|0.0371&nbsp;μs|0.0413&nbsp;μs|0.0135| 0.0000|0.0000|984&nbsp;B|
|#2099|`SendReceive`|net472|2.469&nbsp;μs|0.0476&nbsp;μs|0.0467&nbsp;μs|0.1550| 0.0000|0.0000|987&nbsp;B|
|#2099|`SendReceive`|netcoreapp3.1|1.954&nbsp;μs|0.0245&nbsp;μs|0.0229&nbsp;μs|0.0138| 0.0000|0.0000|984&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #2099

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SerilogBenchmark.EnrichedLog&#8209;net472 | 1.131 | 7,609.02 | 8,604.56 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|net472|7.629&nbsp;μs|0.1523&nbsp;μs|0.1629&nbsp;μs|0.4467| 0.0000|0.0000|2.8&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`EnrichedLog`|netcoreapp3.1|6.976&nbsp;μs|0.1161&nbsp;μs|0.1808&nbsp;μs|0.0377| 0.0000|0.0000|2.61&nbsp;KB|
|#2099|`EnrichedLog`|net472|8.642&nbsp;μs|0.1295&nbsp;μs|0.1211&nbsp;μs|0.4464| 0.0000|0.0000|2.8&nbsp;KB|
|#2099|`EnrichedLog`|netcoreapp3.1|7.066&nbsp;μs|0.1366&nbsp;μs|0.1573&nbsp;μs|0.0350| 0.0000|0.0000|2.61&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean |    Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`StartFinishSpan`|net472|893.5&nbsp;ns|15.20&nbsp;ns|14.22&nbsp;ns|0.0678| 0.0000|0.0000|433&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`StartFinishSpan`|netcoreapp3.1|863.9&nbsp;ns|9.10&nbsp;ns|8.51&nbsp;ns|0.0060| 0.0000|0.0000|432&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`StartFinishScope`|net472|1,151.3&nbsp;ns|19.53&nbsp;ns|18.27&nbsp;ns|0.0807| 0.0000|0.0000|514&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/f2cbb4ba933e8a5be3a4af0fdf6e5b3d74ce49d9)|`StartFinishScope`|netcoreapp3.1|1,004.5&nbsp;ns|12.44&nbsp;ns|11.03&nbsp;ns|0.0079| 0.0000|0.0000|552&nbsp;B|
|#2099|`StartFinishSpan`|net472|902.5&nbsp;ns|17.78&nbsp;ns|23.12&nbsp;ns|0.0678| 0.0000|0.0000|433&nbsp;B|
|#2099|`StartFinishSpan`|netcoreapp3.1|856.2&nbsp;ns|17.13&nbsp;ns|16.02&nbsp;ns|0.0060| 0.0000|0.0000|432&nbsp;B|
|#2099|`StartFinishScope`|net472|1,076.6&nbsp;ns|21.07&nbsp;ns|25.88&nbsp;ns|0.0807| 0.0000|0.0000|514&nbsp;B|
|#2099|`StartFinishScope`|netcoreapp3.1|1,039.2&nbsp;ns|18.47&nbsp;ns|19.76&nbsp;ns|0.0077| 0.0000|0.0000|552&nbsp;B|
</details>

**Review by colin-higgins — APPROVED** [review]

**Pull request merged**